File: DanceCalc/m4d/Models/SongProperty.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations.Schema;
using System.Diagnostics;
using System.Threading;


namespace m4d.Models
{
    public class SongProperty : DbObject
    {
        // Name Syntax: [+-]BaseName[:idx[:qual]]
        // Where default is repalce, + is add, - is delete
        // idx is zeros based indes for multi-value fields (only album at this point?)
        // qual is a qualifier for purchase type (may generalize?)
        //
        // Not implementing this yet, but for artist might allow artist type after the colon
        
        public SongProperty()
        {
        }

        public SongProperty(int songId, string name, string value)
        {
            SongId = songId;
            Name = name;
            Value = value;
        }
        public SongProperty(int songId, string baseName, string value=null, int index = -1, string qual = null)
        {
            SongId = songId;

            string name = null;

            if (index >= 0)
            {
                name = string.Format("{0}:{1:2d}",name,index);
            }

            if (qual != null)
            {
                name = name + ":" + qual;
            }

            Name = name;
            Value = value;
        }

        public Int64 Id { get; set; }
        public int SongId { get; set; }
        public virtual Song Song { get; set; }
        public string Name { get; set; }
        public string Value { get; set; }
        public object ObjectValue
        {
            get
            {
                object ret = null;
                switch (BaseName)
                {
                    case DanceMusicContext.TempoField:
                        // decimal
                        if (Value != null)
                        {
                            decimal v;
                            decimal.TryParse(Value, out v);
                            ret = v;
                        }
                        break;
                    case DanceMusicContext.LengthField:
                    case DanceMusicContext.TrackField:
                    case DanceMusicContext.DanceRatingField:
                        //int
                        if (Value != null)
                        {
                            int v;
                            int.TryParse(Value, out v);
                            ret = v;
                        }
                        break;
                    case DanceMusicContext.TimeField:
                        // time
                        break;

                    default:
                        ret = Value;
                        break;
                }

                return ret;
            }
        }
        public bool IsComplex 
        {
            get { return IsComplexName(Name); }
        }
        public bool IsAction
        {
            get { return IsActionName(Name); }
        }

        public static bool IsComplexName(string name) 
        { 
            return name.Contains(":");
        }
        public static bool IsActionName(string name)
        {
            return name.StartsWith(".");
        }

        public string BaseName
        {
            get
            {
                string baseName = Name;

                // TODO: Deprecate +/-
                if (baseName.StartsWith("+")|| baseName.StartsWith("-"))
                {
                    baseName = baseName.Substring(1);
                }

                int i = baseName.IndexOf(':');

                if (i >= 0)
                {
                    baseName = baseName.Substring(0, i);
                }

                return baseName;
            }
        }

        public int Index
        {
            get
            {
                int idx = 0;

                if (Name.Contains(":"))
                {
                    string[] parts = Name.Split(new char[] { ':' });

                    if (parts.Length > 1)
                    {
                        int.TryParse(parts[1], out idx);
                    }
                }                

                return idx;
            }
        }

        public string Qualifier
        {
            get
            {
                string qual = null;

                if (Name.Contains(":"))
                {
                    string[] parts = Name.Split(new char[] { ':' });

                    if (parts.Length > 2)
                    {
                        qual = parts[2];
                    }
                }

                return qual;
            }
        }

        public static string FormatName(string baseName, int? idx = null, string qualifier = null)
        {
            string name = baseName;

            if (idx != null)
            {
                name += ":" + idx.ToString();
            }

            if (qualifier != null)
            {
                name += ":" + qualifier;
            }

            return name;
        }

        public override void Dump()
        {
            base.Dump();

            string output = string.Format("Id={0},SongId={1},Name={2},Value={3}", Id, SongId, Name, Value);
            Debug.WriteLine(output);
        }
    }
}

File: m4d/Client/src/model/SongProperty.ts
import "reflect-metadata";
import format from "date-fns/format";
import { jsonMember, jsonObject } from "typedjson";

// Name Syntax:
//   BaseName[:idx[:qual]]
//     idx is zeros based index for multi-value fields (only album at this point?)
//     qual is a qualifier for purchase type (may generalize?)
//   Tag(+|-)[DanceQualifier]

export type PropertyValue = string | number | Date | boolean | undefined;

export enum PropertyType {
  // Field names - note that these must be kept in sync with the actual property names
  userField = "User",
  timeField = "Time",
  titleField = "Title",
  artistField = "Artist",
  tempoField = "Tempo",
  lengthField = "Length",
  sampleField = "Sample",
  danceabilityField = "Danceability",
  energyField = "Energy",
  valenceFiled = "Valence",
  titleHashField = "TitleHash",

  // Album Fields
  albumField = "Album",
  publisherField = "Publisher",
  trackField = "Track",
  purchaseField = "Purchase",
  albumListField = "AlbumList",
  albumPromote = "PromoteAlbum",
  albumOrder = "OrderAlbums",

  // Dance Rating
  danceRatingField = "DanceRating",

  // Tags
  addedTags = "Tag+",
  removedTags = "Tag-",

  // User/Song info
  ownerHash = "OwnerHash",
  likeTag = "Like",

  // Proxy Fields
  userProxy = "UserProxy",

  // Curator Fields
  deleteTag = "DeleteTag",

  // Azure Search Fields
  songIdField = "SongId",
  altIdField = "AlternateIds",
  moodField = "Mood",
  beatField = "Beat",
  albumsField = "Albums",
  createdField = "Created",
  editedField = "Edited",
  modifiedField = "Modified",
  dancesField = "Dances",
  usersField = "Users",
  danceTagsInferred = "DanceTagsInferred",
  genreTags = "GenreTags",
  tempoTags = "TempoTags",
  styleTags = "StyleTags",
  otherTags = "OtherTags",
  propertiesField = "Properties",
  serviceIds = "ServiceIds",
  lookupStatus = "LookupStatus",

  // Special cases for reading scraped data
  titleArtistCell = "TitleArtist",
  dancersCell = "Dancers",
  danceTags = "DanceTags",
  songTags = "SongTags",
  measureTempo = "MPM",
  multiDance = "MultiDance",

  // Why does this exist?
  songId = "SongId",

  // Commands
  createCommand = ".Create",
  editCommand = ".Edit",
  deleteCommand = ".Delete",
  mergeCommand = ".Merge",
  undoCommand = ".Undo",
  redoCommand = ".Redo",
  failedLookup = ".FailedLookup",
  noSongId = ".NoSongId", // Pseudo action for serialization
  serializeDeleted = ".SerializeDeleted", // Pseudo action for serialization

  successResult = ".Success",
  failResult = ".Fail",
  messageData = ".Message",
}

@jsonObject
export class SongProperty {
  @jsonMember public name!: string;
  @jsonMember public value!: string;

  public constructor(init?: Partial<SongProperty>) {
    Object.assign(this, init);
  }

  public get valueTyped(): PropertyValue {
    const value = this.value;
    switch (this.baseName) {
      // decimal & float
      case PropertyType.tempoField:
      case PropertyType.danceabilityField:
      case PropertyType.valenceFiled:
      case PropertyType.energyField:
        return this.floatValue;

      // int
      case PropertyType.lengthField:
      case PropertyType.trackField:
      case PropertyType.ownerHash:
        return this.intValue;

      // date
      case PropertyType.timeField:
        return this.dateValue;

      case PropertyType.likeTag:
        return this.booleanValue;

      default:
        return value;
    }
  }

  private get floatValue(): number | undefined {
    const n = Number.parseFloat(this.value);
    return Number.isNaN(n) ? undefined : n;
  }

  private get intValue(): number {
    return Number.parseInt(this.value, 10);
  }

  private get dateValue(): Date {
    // TODO:  Need to standardize on storing UTC time on the server (and in properties)
    return new Date(this.value);
  }

  private get booleanValue(): boolean | undefined {
    const value = this.value;
    if (value.toLowerCase() === "true") {
      return true;
    }
    if (value.toLowerCase() === "false") {
      return false;
    }
    return undefined;
  }

  public get baseName(): string {
    return this.parsePart(0)!;
  }

  public get hasIndex(): boolean {
    return this.hasPart(1);
  }

  public get index(): number {
    const part = this.parsePart(1);
    if (part === undefined) {
      throw new Error(
        `Attempted to retrieve part ${1} from '${
          this.name
        }' which doesn't exist`
      );
    }

    const index = Number.parseInt(part, 10);
    if (Number.isNaN(index)) {
      throw new Error(`Index must be a number, not '${index}'`);
    }

    if (index < 0) {
      throw new Error(`Index must be a postitive integer, not '${index}'`);
    }

    return index;
  }

  public get danceQualifier(): string | undefined {
    return this.parsePart(1);
  }

  public get qualifier(): string | undefined {
    return this.parsePart(2);
  }

  public get isAction(): boolean {
    return this.name.startsWith(".");
  }

  public toString(): string {
    return `${this.name}=${this.value}`;
  }

  public static formatDate(date: Date): string {
    return format(date, "dd-MMM-yyyy hh:mm:ss a");
  }

  private hasPart(index: number): boolean {
    if (index === 0) {
      return true;
    }
    const first = this.name.indexOf(":");
    if (index === 1) {
      return first !== -1;
    }
    if (index === 2) {
      return first !== -1 && first !== this.name.lastIndexOf(":");
    }
    return false;
  }

  private parsePart(index: number): string | undefined {
    const parts = this.name.split(":");
    return parts.length > index ? parts[index] : undefined;
  }
}
